core/kernel/warn_spec.rb

```diff
@@ -76,4 +76,53 @@
       warn("")
     }.should output(nil, /\n/)
   end
+
+  ruby_version_is "2.5" do
+    describe ":uplevel keyword argument" do
+      it "prepends a message with specified line from the backtrace" do
+        w = KernelSpecs::WarnInNestedCall.new
+
+        -> { w.f4("foo", 0) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 1) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f1_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 2) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f2_call_lineno}: warning: foo|)
+        -> { w.f4("foo", 3) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.f3_call_lineno}: warning: foo|)
+      end
+
+      it "does not prepend anything if line number is too big" do
+        w = KernelSpecs::WarnInNestedCall.new
+        -> { w.f4("foo", 100) }.should output(nil, "foo\n")
+      end
+
+      it "prepends even if a message is empty or nil" do
+        w = KernelSpecs::WarnInNestedCall.new
+
+        -> { w.f4("", 0) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: \n$|)
+        -> { w.f4(nil, 0) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: \n$|)
+      end
+
+      it "converts value to Integer" do
+        w = KernelSpecs::WarnInNestedCall.new
+
+        -> { w.f4(0.1) }.should output(nil, %r|classes.rb:#{w.warn_call_lineno}:|)
+        -> { w.f4(Rational(1, 2)) }.should output(nil, %r|classes.rb:#{w.warn_call_lineno}:|)
+      end
+
+      it "raises ArgumentError if passed negative value" do
+        -> { warn "", uplevel: -2 }.should raise_error(ArgumentError)
+        -> { warn "", uplevel: -100 }.should raise_error(ArgumentError)
+      end
+
+      it "processes value -1 like 0" do
+        w = KernelSpecs::WarnInNestedCall.new
+        -> { w.f4("foo", -1) }.should output(nil, %r|core/kernel/fixtures/classes.rb:#{w.warn_call_lineno}: warning: foo|)
+      end
```

Comment:
@eregon I would say it's a bug

Reply:
Sorry, I missed this message until now (I was quite busy at RubyKaigi).
What do you think should be the behavior? ArgumentError like above?

Reply:
This was fixed in MRI trunk, it raises an ArgumentError now:
```
$ ruby -e 'warn "", uplevel: -1'
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `warn': negative level (-1) (ArgumentError)
```

Reply:
I filed https://bugs.ruby-lang.org/issues/14846 to track the backport to 2.5.x.